relay-sampling/src/condition.rs

```diff
@@ -0,0 +1,414 @@
+}
+
+/// Combines multiple conditions using logical OR.
+///
+/// This condition matches if **any** of the inner conditions matches.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct OrCondition {
+    /// Inner rules to combine.
+    pub inner: Vec<RuleCondition>,
+}
+
+impl OrCondition {
+    fn supported(&self) -> bool {
+        self.inner.iter().all(RuleCondition::supported)
+    }
+
+    fn matches<T>(&self, value: &T) -> bool
+    where
+        T: Getter,
+    {
+        self.inner.iter().any(|cond| cond.matches(value))
+    }
+}
+
+/// Combines multiple conditions using logical AND.
+///
+/// This condition matches if **all** of the inner conditions matches.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct AndCondition {
+    /// Inner rules to combine.
+    pub inner: Vec<RuleCondition>,
+}
+
+impl AndCondition {
+    fn supported(&self) -> bool {
+        self.inner.iter().all(RuleCondition::supported)
+    }
+    fn matches<T>(&self, value: &T) -> bool
+    where
+        T: Getter,
+    {
+        self.inner.iter().all(|cond| cond.matches(value))
+    }
+}
+
+/// Applies logical NOT to a condition.
+///
+/// This condition matches if the inner condition does not match.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct NotCondition {
+    /// An inner rule to negate.
+    pub inner: Box<RuleCondition>,
+}
+
+impl NotCondition {
+    fn supported(&self) -> bool {
+        self.inner.supported()
+    }
+
+    fn matches<T>(&self, value: &T) -> bool
+    where
+        T: Getter,
+    {
+        !self.inner.matches(value)
+    }
+}
+
+/// A condition from a sampling rule.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "op")]
+pub enum RuleCondition {
+    /// A condition that compares values for equality.
+    Eq(EqCondition),
+    /// A condition that applies `>=`.
+    Gte(GteCondition),
+    /// A condition that applies `<=`.
+    Lte(LteCondition),
+    /// A condition that applies `>`.
+    Gt(GtCondition),
+    /// A condition that applies `<`.
+    Lt(LtCondition),
+    /// A condition that uses glob matching.
+    Glob(GlobCondition),
+    /// Combines multiple conditions using logical OR.
+    Or(OrCondition),
+    /// Combines multiple conditions using logical AND.
+    And(AndCondition),
+    /// Applies logical NOT to a condition.
+    Not(NotCondition),
+    /// An unsupported condition for future compatibility.
+    #[serde(other)]
+    Unsupported,
+}
+
+impl RuleCondition {
+    /// Returns a condition that matches everything.
+    pub fn all() -> Self {
+        Self::And(AndCondition { inner: Vec::new() })
+    }
+
+    /// Checks if Relay supports this condition (in other words if the condition had any unknown configuration
+    /// which was serialized as "Unsupported" (because the configuration is either faulty or was created for a
+    /// newer relay that supports some other condition types)
```

Comment:
```suggestion
    /// Checks if Relay supports this condition. In other words, if the condition had any unknown configuration
    /// which was serialized as "Unsupported", because the configuration is either faulty or was created for a
    /// newer relay that supports some other condition types.
```
? (I find two nested parenthesis hard to follow)